# 前言
- 这⾥我们使⽤的是1984年美国国会投票的数据，这个数据集中包含了每个美国众议院议员对CQA确定的16个主要选票的投票。CQA列出了九种不同类型的投票：投票赞成、配对赞成和宣布赞成（这三个简化为yea)，投票反对、配对反对、宣布反对（这三个简化为nay）, 投票、为避免利益冲突⽽投票、弃权或以其他⽅式表明⽴场（这三个简化为未知意向）
- 数据集来源于UCI数据库：http://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records

#  导入数据集并进行简单数据处理

In [1]:
import pandas as pd

In [2]:
votes_ = pd.read_csv('house-votes-84.data.txt',sep=',',header=None)
votes_.head()

0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15 16
0  republican  n  y  n  y  y  y  n  n  n  y  ?  y  y  y  n  y
1  republican  n  y  n  y  y  y  n  n  n  n  n  y  y  y  n  ?
2    democrat  ?  y  y  ?  y  y  n  n  n  n  y  n  y  y  n  n
3    democrat  n  y  y  n  ?  y  n  n  n  n  y  n  y  n  n  y
4    democrat  y  y  y  n  y  y  n  n  n  n  y  ?  y  y  y  y

In [3]:
votes_.columns = ['党派'
 ,'残疾⼈婴幼⼉提案'
 ,'⽔项⽬费⽤分摊'
 ,'预算决议案'
 ,'医⽣费⽤冻结决议案'
 ,'萨尔瓦多援助'
 ,'校园宗教团体决议'
 ,'反卫星禁试决议'
 ,'援助尼加拉⽠反政府'
 ,'MX导弹决议案'
 ,'移⺠决议案'
 ,'合成燃料公司削减决议'
 ,'教育⽀出决议'
 ,'超级基⾦起诉权'
 ,'犯罪决议案'
 ,'免税出⼝决议案'
 ,'南⾮出⼝管理决议案']

In [4]:
votes_

党派 残疾⼈婴幼⼉提案 ⽔项⽬费⽤分摊 预算决议案 医⽣费⽤冻结决议案 萨尔瓦多援助 校园宗教团体决议 反卫星禁试决议  \
0    republican        n       y     n         y      y        y       n   
1    republican        n       y     n         y      y        y       n   
2      democrat        ?       y     y         ?      y        y       n   
3      democrat        n       y     y         n      ?        y       n   
4      democrat        y       y     y         n      y        y       n   
..          ...      ...     ...   ...       ...    ...      ...     ...   
430  republican        n       n     y         y      y        y       n   
431    democrat        n       n     y         n      n        n       y   
432  republican        n       ?     n         y      y        y       n   
433  republican        n       n     n         y      y        y       ?   
434  republican        n       y     n         y      y        y       n   

    援助尼加拉⽠反政府 MX导弹决议案 移⺠决议案 合成燃料公司削减决议 教育⽀出决议 超级基⾦起诉权 犯罪决议案 免税出⼝决议案 南⾮出⼝管理决议案  
0           n       n     y          ?      y       y     y       n         y  
1           n       n     n          n      y       y     y       n         ?  
2           n       n     n          y      n       y     y       n         n  
3           n       n     n          y      n       y     n       n         y  
4           n       n     n          y      ?       y     y       y         y  
..        ...     ...   ...        ...    ...     ...   ...     ...       ...  
430         n       y     y          n      y       y     y       n         y  
431         y       y     y          n      n       n     n       n         y  
432         n       n     n          y      y       y     y       n         y  
433         ?       ?     ?          n      y       y     y       n         y  
434         n       n     y          n      y       y     y       ?         n  

[435 rows x 17 columns]

In [5]:
# 将列名和投票结果合并
for i in range(1,votes_.shape[1]): #从1开始是因为不对党派进⾏操作
     votes_.iloc[:,i] = votes_.iloc[:,i].apply(lambda x:votes_.columns[i]+'-'+x)

In [6]:
votes_

党派    残疾⼈婴幼⼉提案    ⽔项⽬费⽤分摊    预算决议案    医⽣费⽤冻结决议案    萨尔瓦多援助  \
0    republican  残疾⼈婴幼⼉提案-n  ⽔项⽬费⽤分摊-y  预算决议案-n  医⽣费⽤冻结决议案-y  萨尔瓦多援助-y   
1    republican  残疾⼈婴幼⼉提案-n  ⽔项⽬费⽤分摊-y  预算决议案-n  医⽣费⽤冻结决议案-y  萨尔瓦多援助-y   
2      democrat  残疾⼈婴幼⼉提案-?  ⽔项⽬费⽤分摊-y  预算决议案-y  医⽣费⽤冻结决议案-?  萨尔瓦多援助-y   
3      democrat  残疾⼈婴幼⼉提案-n  ⽔项⽬费⽤分摊-y  预算决议案-y  医⽣费⽤冻结决议案-n  萨尔瓦多援助-?   
4      democrat  残疾⼈婴幼⼉提案-y  ⽔项⽬费⽤分摊-y  预算决议案-y  医⽣费⽤冻结决议案-n  萨尔瓦多援助-y   
..          ...         ...        ...      ...          ...       ...   
430  republican  残疾⼈婴幼⼉提案-n  ⽔项⽬费⽤分摊-n  预算决议案-y  医⽣费⽤冻结决议案-y  萨尔瓦多援助-y   
431    democrat  残疾⼈婴幼⼉提案-n  ⽔项⽬费⽤分摊-n  预算决议案-y  医⽣费⽤冻结决议案-n  萨尔瓦多援助-n   
432  republican  残疾⼈婴幼⼉提案-n  ⽔项⽬费⽤分摊-?  预算决议案-n  医⽣费⽤冻结决议案-y  萨尔瓦多援助-y   
433  republican  残疾⼈婴幼⼉提案-n  ⽔项⽬费⽤分摊-n  预算决议案-n  医⽣费⽤冻结决议案-y  萨尔瓦多援助-y   
434  republican  残疾⼈婴幼⼉提案-n  ⽔项⽬费⽤分摊-y  预算决议案-n  医⽣费⽤冻结决议案-y  萨尔瓦多援助-y   

       校园宗教团体决议    反卫星禁试决议    援助尼加拉⽠反政府    MX导弹决议案    移⺠决议案    合成燃料公司削减决议  \
0    校园宗教团体决议-y  反卫星禁试决议-n  援助尼加拉⽠反政府-n  MX导弹决议案-n  移⺠决议案-y  合成燃料公司削减决议-?   
1    校园宗教团体决议-y  反卫星禁试决议-n  援助尼加拉⽠反政府-n  MX导弹决议案-n  移⺠决议案-n  合成燃料公司削减决议-n   
2    校园宗教团体决议-y  反卫星禁试决议-n  援助尼加拉⽠反政府-n  MX导弹决议案-n  移⺠决议案-n  合成燃料公司削减决议-y   
3    校园宗教团体决议-y  反卫星禁试决议-n  援助尼加拉⽠反政府-n  MX导弹决议案-n  移⺠决议案-n  合成燃料公司削减决议-y   
4    校园宗教团体决议-y  反卫星禁试决议-n  援助尼加拉⽠反政府-n  MX导弹决议案-n  移⺠决议案-n  合成燃料公司削减决议-y   
..          ...        ...          ...        ...      ...           ...   
430  校园宗教团体决议-y  反卫星禁试决议-n  援助尼加拉⽠反政府-n  MX导弹决议案-y  移⺠决议案-y  合成燃料公司削减决议-n   
431  校园宗教团体决议-n  反卫星禁试决议-y  援助尼加拉⽠反政府-y  MX导弹决议案-y  移⺠决议案-y  合成燃料公司削减决议-n   
432  校园宗教团体决议-y  反卫星禁试决议-n  援助尼加拉⽠反政府-n  MX导弹决议案-n  移⺠决议案-n  合成燃料公司削减决议-y   
433  校园宗教团体决议-y  反卫星禁试决议-?  援助尼加拉⽠反政府-?  MX导弹决议案-?  移⺠决议案-?  合成燃料公司削减决议-n   
434  校园宗教团体决议-y  反卫星禁试决议-n  援助尼加拉⽠反政府-n  MX导弹决议案-n  移⺠决议案-y  合成燃料公司削减决议-n   

       教育⽀出决议    超级基⾦起诉权    犯罪决议案    免税出⼝决议案    南⾮出⼝管理决议案  
0    教育⽀出决议-y  超级基⾦起诉权-y  犯罪决议案-y  免税出⼝决议案-n  南⾮出⼝管理决议案-y  
1    教育⽀出决议-y  超级基⾦起诉权-y  犯罪决议案-y  免税出⼝决议案-n  南⾮出⼝管理决议案-?  
2    教育⽀出决议-n  超级基⾦起诉权-y  犯罪决议案-y  免税出⼝决议案-n  南⾮出⼝管理决议案-n  
3    教育⽀出决议-n  超级基⾦起诉权-y  犯罪决议案-n  免税出⼝决议案-n  南⾮出⼝管理决议案-y  
4    教育⽀出决议-?  超级基⾦起诉权-y  犯罪决议案-y  免税出⼝决议案-y  南⾮出⼝管理决议案-y  
..        ...        ...      ...        ...          ...  
430  教育⽀出决议-y  超级基⾦起诉权-y  犯罪决议案-y  免税出⼝决议案-n  南⾮出⼝管理决议案-y  
431  教育⽀出决议-n  超级基⾦起诉权-n  犯罪决议案-n  免税出⼝决议案-n  南⾮出⼝管理决议案-y  
432  教育⽀出决议-y  超级基⾦起诉权-y  犯罪决议案-y  免税出⼝决议案-n  南⾮出⼝管理决议案-y  
433  教育⽀出决议-y  超级基⾦起诉权-y  犯罪决议案-y  免税出⼝决议案-n  南⾮出⼝管理决议案-y  
434  教育⽀出决议-y  超级基⾦起诉权-y  犯罪决议案-y  免税出⼝决议案-?  南⾮出⼝管理决议案-n  

[435 rows x 17 columns]

In [7]:
# 将数据变为列表
voteslist=votes_.values.tolist()
voteslist

[['republican',
  '残疾⼈婴幼⼉提案-n',
  '⽔项⽬费⽤分摊-y',
  '预算决议案-n',
  '医⽣费⽤冻结决议案-y',
  '萨尔瓦多援助-y',
  '校园宗教团体决议-y',
  '反卫星禁试决议-n',
  '援助尼加拉⽠反政府-n',
  'MX导弹决议案-n',
  '移⺠决议案-y',
  '合成燃料公司削减决议-?',
  '教育⽀出决议-y',
  '超级基⾦起诉权-y',
  '犯罪决议案-y',
  '免税出⼝决议案-n',
  '南⾮出⼝管理决议案-y'],
 ['republican',
  '残疾⼈婴幼⼉提案-n',
  '⽔项⽬费⽤分摊-y',
  '预算决议案-n',
  '医⽣费⽤冻结决议案-y',
  '萨尔瓦多援助-y',
  '校园宗教团体决议-y',
  '反卫星禁试决议-n',
  '援助尼加拉⽠反政府-n',
  'MX导弹决议案-n',
  '移⺠决议案-n',
  '合成燃料公司削减决议-n',
  '教育⽀出决议-y',
  '超级基⾦起诉权-y',
  '犯罪决议案-y',
  '免税出⼝决议案-n',
  '南⾮出⼝管理决议案-?'],
 ['democrat',
  '残疾⼈婴幼⼉提案-?',
  '⽔项⽬费⽤分摊-y',
  '预算决议案-y',
  '医⽣费⽤冻结决议案-?',
  '萨尔瓦多援助-y',
  '校园宗教团体决议-y',
  '反卫星禁试决议-n',
  '援助尼加拉⽠反政府-n',
  'MX导弹决议案-n',
  '移⺠决议案-n',
  '合成燃料公司削减决议-y',
  '教育⽀出决议-n',
  '超级基⾦起诉权-y',
  '犯罪决议案-y',
  '免税出⼝决议案-n',
  '南⾮出⼝管理决议案-n'],
 ['democrat',
  '残疾⼈婴幼⼉提案-n',
  '⽔项⽬费⽤分摊-y',
  '预算决议案-y',
  '医⽣费⽤冻结决议案-n',
  '萨尔瓦多援助-?',
  '校园宗教团体决议-y',
  '反卫星禁试决议-n',
  '援助尼加拉⽠反政府-n',
  'MX导弹决议案-n',
  '移⺠决议案-n',
  '合成燃料公司削减决议-y',
  '教育⽀出决议-n',
  '超级基⾦起诉权-y',
  '犯罪决议案-n

#  挖掘所有的频繁项集

 - Apriori算法来进⾏频繁项集的挖掘，这⾥设定默认的⽀持度阈值50%。

In [8]:
%run Apriori.ipynb

In [9]:
L,suppData = apriori(voteslist, min_support = 0.5)

In [10]:
L

[[frozenset({'免税出⼝决议案-n'}),
  frozenset({'南⾮出⼝管理决议案-y'}),
  frozenset({'校园宗教团体决议-y'}),
  frozenset({'残疾⼈婴幼⼉提案-n'}),
  frozenset({'犯罪决议案-y'}),
  frozenset({'合成燃料公司削减决议-n'}),
  frozenset({'democrat'}),
  frozenset({'教育⽀出决议-n'}),
  frozenset({'预算决议案-y'}),
  frozenset({'医⽣费⽤冻结决议案-n'}),
  frozenset({'反卫星禁试决议-y'}),
  frozenset({'援助尼加拉⽠反政府-y'})],
 [frozenset({'democrat', '预算决议案-y'}),
  frozenset({'democrat', '医⽣费⽤冻结决议案-n'}),
  frozenset({'医⽣费⽤冻结决议案-n', '预算决议案-y'}),
  frozenset({'democrat', '援助尼加拉⽠反政府-y'})],
 [frozenset({'democrat', '医⽣费⽤冻结决议案-n', '预算决议案-y'})]]

In [11]:
df = pd.DataFrame()
for li in L:
    for i in li:
        dic = {}
        dic['频繁项集'] = str([item for item in i])
        row = pd.DataFrame(dic, index=[0])
        df = pd.concat([df,row],ignore_index=True)
df

频繁项集
0                            ['免税出⼝决议案-n']
1                          ['南⾮出⼝管理决议案-y']
2                           ['校园宗教团体决议-y']
3                           ['残疾⼈婴幼⼉提案-n']
4                              ['犯罪决议案-y']
5                         ['合成燃料公司削减决议-n']
6                             ['democrat']
7                             ['教育⽀出决议-n']
8                              ['预算决议案-y']
9                          ['医⽣费⽤冻结决议案-n']
10                           ['反卫星禁试决议-y']
11                         ['援助尼加拉⽠反政府-y']
12                 ['预算决议案-y', 'democrat']
13             ['医⽣费⽤冻结决议案-n', 'democrat']
14              ['预算决议案-y', '医⽣费⽤冻结决议案-n']
15             ['援助尼加拉⽠反政府-y', 'democrat']
16  ['预算决议案-y', '医⽣费⽤冻结决议案-n', 'democrat']

In [12]:
suppData

{frozenset({'MX导弹决议案-n'}): 0.4735632183908046,
 frozenset({'republican'}): 0.38620689655172413,
 frozenset({'⽔项⽬费⽤分摊-y'}): 0.4482758620689655,
 frozenset({'免税出⼝决议案-n'}): 0.535632183908046,
 frozenset({'医⽣费⽤冻结决议案-y'}): 0.4068965517241379,
 frozenset({'南⾮出⼝管理决议案-y'}): 0.6183908045977011,
 frozenset({'反卫星禁试决议-n'}): 0.41839080459770117,
 frozenset({'合成燃料公司削减决议-?'}): 0.04827586206896552,
 frozenset({'援助尼加拉⽠反政府-n'}): 0.4091954022988506,
 frozenset({'教育⽀出决议-y'}): 0.3931034482758621,
 frozenset({'校园宗教团体决议-y'}): 0.6252873563218391,
 frozenset({'残疾⼈婴幼⼉提案-n'}): 0.542528735632184,
 frozenset({'犯罪决议案-y'}): 0.5701149425287356,
 frozenset({'移⺠决议案-y'}): 0.496551724137931,
 frozenset({'萨尔瓦多援助-y'}): 0.48735632183908045,
 frozenset({'超级基⾦起诉权-y'}): 0.4804597701149425,
 frozenset({'预算决议案-n'}): 0.3931034482758621,
 frozenset({'南⾮出⼝管理决议案-?'}): 0.23908045977011494,
 frozenset({'合成燃料公司削减决议-n'}): 0.6068965517241379,
 frozenset({'移⺠决议案-n'}): 0.48735632183908045,
 frozenset({'democrat'}): 0.6137931034482759,
 fro

In [13]:
df1 = pd.DataFrame()
row = pd.DataFrame(suppData, index=[0])
df1 = pd.concat([df1,row],ignore_index=True)
df1.T

0
(MX导弹决议案-n)                           0.473563
(republican)                          0.386207
(⽔项⽬费⽤分摊-y)                           0.448276
(免税出⼝决议案-n)                           0.535632
(医⽣费⽤冻结决议案-y)                         0.406897
...                                        ...
(反卫星禁试决议-y, 犯罪决议案-y)                  0.190805
(援助尼加拉⽠反政府-y, 犯罪决议案-y)                0.170115
(预算决议案-y, 医⽣费⽤冻结决议案-n, democrat)      0.503448
(预算决议案-y, 援助尼加拉⽠反政府-y, democrat)      0.466667
(医⽣费⽤冻结决议案-n, 援助尼加拉⽠反政府-y, democrat)  0.482759

[119 rows x 1 columns]

#  挖掘强关联规则
- 这⾥设定置信度阈值为90%

In [14]:
brl = gen_rules(L,suppData, min_conf=0.9)

In [15]:
brl

[(frozenset({'预算决议案-y'}),
  '->',
  frozenset({'democrat'}),
  ' = ',
  0.9130434782608695),
 (frozenset({'democrat'}),
  '->',
  frozenset({'医⽣费⽤冻结决议案-n'}),
  ' = ',
  0.9176029962546816),
 (frozenset({'医⽣费⽤冻结决议案-n'}),
  '->',
  frozenset({'democrat'}),
  ' = ',
  0.9919028340080972),
 (frozenset({'援助尼加拉⽠反政府-y'}),
  '->',
  frozenset({'democrat'}),
  ' = ',
  0.9008264462809917),
 (frozenset({'democrat', '预算决议案-y'}),
  '->',
  frozenset({'医⽣费⽤冻结决议案-n'}),
  ' = ',
  0.9480519480519481),
 (frozenset({'医⽣费⽤冻结决议案-n', '预算决议案-y'}),
  '->',
  frozenset({'democrat'}),
  ' = ',
  1.0)]

In [16]:
# '''
# If                                         Then                                           可信度
# y-援助尼加拉⽠反政府                       democrat                                        90.0%
# y-预算决议案                               democrat                                        91.3%
# democrat（⺠主党⼈）                       n-医⽣费⽤冻结决议案                            91.8%
# y-预算决议案，democrat                     n-医⽣费⽤冻结决议案                            94.8%
# n-医⽣费⽤冻结决议案                       democrat                                        99.1%
# y-预算决议案，n-医⽣费⽤冻结决议案         democrat                                        100.0%
# '''

- 这些强规则是在⽀持度为50%的情况下挖掘出来的，也就是说，这些规则⾄少在50%以上的记录中出现过。对于{y-预算决议案，n-医⽣费⽤冻结决议案}→{democrat}这条规则，在100%的情况下是成⽴的，也就是说，这是⼀个必然事件，只要他给预算决议案投了赞成票，并且医⽣费⽤冻结决议案投了反对票，那么这个⼈⼀定是⺠主党。
- 当设置⽀持度阈值为50%，置信度阈值为90%的时候，我们挖掘出来的强关联规则全部是跟⺠主党相关的，为什么会出现这样的情况呢?

In [17]:
votes_['党派'].value_counts()
votes_['党派'].value_counts()/votes_.shape[0]

democrat      0.613793
republican    0.386207
Name: 党派, dtype: float64

- 可以看出，共和党所占⽐例仅为38.6%，如果设定⽀持度为50%，它根本就不会进⼊频繁项集。 如果我们⼀定想要查看共和党的⼀些特征，挖掘与共和党相关的⼀些强关联规则，那么我们可以将⽀持度降低⼀点， ⽐如30%，然后再来进⾏频繁项集的挖掘、关联规则的提取。但是如果降低了⽀持度，那频繁项集的数量⼀定会很⼤程度的增加，置信度的设定就要更⾼⼀点，否则数据量⼤的话很容易卡死

In [18]:
L,suppData = apriori(voteslist, min_support = 0.3) #设定最⼩⽀持度为30%
brl = gen_rules(L, suppData, min_conf=0.99)

In [19]:
brl

[(frozenset({'医⽣费⽤冻结决议案-n'}),
  '->',
  frozenset({'democrat'}),
  ' = ',
  0.9919028340080972),
 (frozenset({'MX导弹决议案-n', 'republican'}),
  '->',
  frozenset({'医⽣费⽤冻结决议案-y'}),
  ' = ',
  0.9931506849315067),
 (frozenset({'MX导弹决议案-n', 'republican'}),
  '->',
  frozenset({'萨尔瓦多援助-y'}),
  ' = ',
  0.9931506849315067),
 (frozenset({'MX导弹决议案-n', '医⽣费⽤冻结决议案-y'}),
  '->',
  frozenset({'萨尔瓦多援助-y'}),
  ' = ',
  0.9935064935064933),
 (frozenset({'republican', '援助尼加拉⽠反政府-n'}),
  '->',
  frozenset({'医⽣费⽤冻结决议案-y'}),
  ' = ',
  0.9924812030075189),
 (frozenset({'republican', '教育⽀出决议-y'}),
  '->',
  frozenset({'医⽣费⽤冻结决议案-y'}),
  ' = ',
  0.9925925925925927),
 (frozenset({'republican', '萨尔瓦多援助-y'}),
  '->',
  frozenset({'医⽣费⽤冻结决议案-y'}),
  ' = ',
  0.9936305732484076),
 (frozenset({'republican', '超级基⾦起诉权-y'}),
  '->',
  frozenset({'医⽣费⽤冻结决议案-y'}),
  ' = ',
  0.9926470588235294),
 (frozenset({'republican', '援助尼加拉⽠反政府-n'}),
  '->',
  frozenset({'萨尔瓦多援助-y'}),
  ' = ',
  0.9924812030075189),
 (frozenset({

In [20]:
df2 = pd.DataFrame()
for li in brl:
    dic = {}
    dic['if'] = str([i for i in li[0]])
    dic['then'] = str([i for i in li[2]])
    dic['conf'] = np.round(li[4], 3)
    row = pd.DataFrame(dic, index=[0])
    df2 = pd.concat([df2,row],ignore_index=True)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [21]:
df2

if  \
0                                      ['医⽣费⽤冻结决议案-n']   
1                          ['republican', 'MX导弹决议案-n']   
2                          ['republican', 'MX导弹决议案-n']   
3                         ['MX导弹决议案-n', '医⽣费⽤冻结决议案-y']   
4                        ['republican', '援助尼加拉⽠反政府-n']   
..                                                 ...   
254  ['反卫星禁试决议-y', '援助尼加拉⽠反政府-y', '犯罪决议案-n', '萨尔瓦多援...   
255  ['反卫星禁试决议-y', '医⽣费⽤冻结决议案-n', '援助尼加拉⽠反政府-y', 'M...   
256  ['反卫星禁试决议-y', '医⽣费⽤冻结决议案-n', '援助尼加拉⽠反政府-y', 'M...   
257  ['反卫星禁试决议-y', '援助尼加拉⽠反政府-y', '萨尔瓦多援助-n', '预算决议...   
258  ['医⽣费⽤冻结决议案-n', '反卫星禁试决议-y', '援助尼加拉⽠反政府-y', '萨...   

                            then   conf  
0                   ['democrat']  0.992  
1                ['医⽣费⽤冻结决议案-y']  0.993  
2                   ['萨尔瓦多援助-y']  0.993  
3                   ['萨尔瓦多援助-y']  0.994  
4                ['医⽣费⽤冻结决议案-y']  0.992  
..                           ...    ...  
254  ['医⽣费⽤冻结决议案-n', 'democrat']  0.992  
255                 ['democrat']  0.992  
256                 ['democrat']  1.000  
257              ['医⽣费⽤冻结决议案-n']  1.000  
258                 ['democrat']  1.000  

[259 rows x 3 columns]

In [22]:
df2.to_excel('relation.xlsx')

 - 我们还可以进行许多进一步探索